pkg/controller/certificaterequests/acme/acme_test.go

```diff
@@ -0,0 +1,211 @@
+			Name:  "acme-issuer",
+			Group: certmanager.GroupName,
+			Kind:  "Issuer",
+		}),
+	)
+
+	_, rsaPEMCert := testcr.GenerateSelfSignedCertFromCR(t, testCR, rsaPK, time.Hour*24*60)
+	testOrder := buildOrderMust(t, testCR, csr)
+
+	tests := map[string]testT{
+		"a badly formed CSR should report failure": {
+			issuer: gen.Issuer("acme-issuer"),
+			certificateRequest: gen.CertificateRequestFrom(testCR,
+				gen.SetCertificateRequestCSR([]byte("a bad csr")),
+			),
+			builder: &testpkg.Builder{
+				KubeObjects:        []runtime.Object{},
+				CertManagerObjects: []runtime.Object{},
+				ExpectedEvents: []string{
+					"Warning ErrorParsingCSR Failed to decode CSR in spec: error decoding certificate request PEM block",
+				},
+				CheckFn: testcr.MustNoResponse,
+			},
+			expectedErr: false,
+		},
+
+		//TODO: Think of a creative way to get `buildOrder` to fail :thinking_face:
+
+		"if order doesn't exist then attempt to create one": {
+			issuer:             gen.Issuer("acme-issuer"),
+			certificateRequest: testCR,
+			builder: &testpkg.Builder{
+				CertManagerObjects: []runtime.Object{},
+				ExpectedActions: []testpkg.Action{
+					testpkg.NewAction(coretesting.NewCreateAction(
+						v1alpha1.SchemeGroupVersion.WithResource("orders"),
+						gen.DefaultTestNamespace,
+						testOrder,
+					)),
+				},
+				ExpectedEvents: []string{
+					`Normal OrderPending Waiting on certificate issuance from order default-unit-test-ns/test-cr-3958469914: order is currently pending: ""`,
+				},
+				CheckFn: testcr.MustNoResponse,
+			},
+			expectedErr: false,
+		},
+
+		"if we fail to get the order resource then we should report pending": {
+			issuer:             gen.Issuer("acme-issuer"),
+			certificateRequest: testCR,
+			builder: &testpkg.Builder{
+				ExpectedEvents: []string{
+					`Normal ErrorGettingOrder Failed to get order resource default-unit-test-ns/test-cr-3958469914: this is a network error`,
+				},
+				CheckFn: testcr.MustNoResponse,
+			},
+			expectedErr: true,
+			FakeOrderLister: &testlisters.FakeOrderLister{
+				OrdersFn: func(namespace string) cmlisters.OrderNamespaceLister {
+					return &testlisters.FakeOrderNamespaceLister{
+						GetFn: func(name string) (ret *v1alpha1.Order, err error) {
+							return nil, errors.New("this is a network error")
+						},
+					}
+				},
+			},
+		},
+
+		"if the order resource is in a failed state then we should report failure": {
+			issuer:             gen.Issuer("acme-issuer"),
+			certificateRequest: testCR,
+			builder: &testpkg.Builder{
+				ExpectedEvents: []string{
+					`Warning OrderFailed Failed to resolve order resource default-unit-test-ns/test-cr-3958469914: order is in failure state "invalid"`,
+				},
+				CertManagerObjects: []runtime.Object{gen.OrderFrom(testOrder,
+					gen.SetOrderState(v1alpha1.Invalid),
+				)},
+				CheckFn: testcr.MustNoResponse,
+			},
+			expectedErr: false,
+		},
+
+		"if the order is in a non failure state or valid, then report pending": {
```

Comment:
`if the order is in a non failure state or valid,` -> `if the order is in an unknown state,`